pkg/util/util.go

```diff
@@ -0,0 +1,74 @@
+package util
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"time"
+)
+
+// This function validates the URL, then tries to fetch the data with retries
+// and then reads and returns the data as []byte
+// Returns an error if the URL is invalid, or fetching the data failed or
+// if reading the response body fails.
+func GetURLData(urlString string, attempts int) ([]byte, error) {
+	// Validate URL
+	_, err := url.ParseRequestURI(urlString)
+	if err != nil {
+		return nil, fmt.Errorf("invalid URL: %v", urlString)
+	}
+
+	// Fetch the URL and store the response body
+	body, err := FetchURLWithRetries(urlString, attempts, time.Second)
+	if err != nil {
+		return nil, fmt.Errorf("failed fetching data from the URL %v: %s", urlString, err)
+	}
+	defer body.Close()
+
+	// Read from the response body, ioutil.ReadAll will return []byte
+	data, err := ioutil.ReadAll(body)
+	if err != nil {
+		return nil, fmt.Errorf("reading from response body failed: %s", err)
+	}
+
+	return data, nil
+}
+
+// Try to fetch the given url string, and make <attempts> attempts at it.
+// Wait for <duration> time between each try.
+// This returns the response body upon successful fetch
+// The passed URL is not validated, so validate the URL before passing to this
+// function
+func FetchURLWithRetries(url string, attempts int, duration time.Duration) (io.ReadCloser, error) {
+	var body io.ReadCloser
+	var err error
+
+	for i := 0; i < attempts; i++ {
+		var response *http.Response
+
+		// sleep for <duration> seconds before trying again
+		if i > 0 {
+			time.Sleep(duration)
+		}
+
+		// retry if http.Get fails
+		// if all the retries fail, then return statement at the end of the
+		// function will return this err received from http.Get
+		response, err = http.Get(url)
+		if err != nil {
+			continue
+		}
+
+		// if the status code is not 200 OK, return an error
+		if response.StatusCode != http.StatusOK {
+			return nil, fmt.Errorf("unable to fetch %v, server returned status code %v", url, response.StatusCode)
```

Comment:
> If 3 request would hit e.g. 404, 404, 200 this condition would end up with the first one and didn't continue, right? This may be because of cache propagation. I would consider retrying in that case. (citing the 404 description: "The requested resource could not be found but may be available in the future. Subsequent requests by the client are permissible.")

Yep, if a 404 is encountered in the first time, no retries are made and opencompose errors out.
The cache propagation case you have mentioned seems like an edge case, should we really take that into consideration in opencompose?